### PR TITLE
[WIP] adds contributor guides to the docs

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This takes @bbbales2 [#192](https://github.com/stan-dev/example-models/pull/192) PR and adds it to the stan math docs. I broke things up a bit into separate sections and fleshed out an example based on Ben's stuff. I like the example because it's pretty much real Stan math code and I think makes it easier to see the context on how we use the special widgets etc. we have. I'd like to add another example that's a bit more complicated for something like `multiply` but I'll do that later. This still needs some heavy editing and I'll tag people for particular sections.

This also changes the settings in `doxygen.cfg` so that even if there are warnings the docs are still built. I think it's annoying that some versions of doxygen compile the docs fine while others don't. This is most entirely because of warnings like 

```
stan/math/rev/functor/idas_integrator.hpp:63: warning: argument 'dae' from the argument list of stan::math::idas_integrator::solve has multiple @param documentation sections
```

Which comes from us having a lot of different specializations and doxygen having trouble working out the docs for each. So I set the `WARN_AS_ERROR` flag for doxygen to `NO`

```
# If the WARN_AS_ERROR tag is set to YES then doxygen will immediately stop when
# a warning is encountered.
# The default value is: NO.

WARN_AS_ERROR          = NO
```

## Release notes

Adds docs for new contributors 

## Checklist

- [x] Math issue #(issue number)

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
You can make the docs locally by running `make doxygen`

**Comment by bbbales2** [comment]
Nice nice! Yeah gimme bite size pieces and happy to add on.

File: doxygen/contributor_help_pages/common_pitfalls.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+```cpp
+template <typename T,
+          require_eigen_t<T>* = nullptr>
+auto myfunc(const T& x) {
+  const auto& x_ref = to_ref(x);
+}
+```
+
+### Holders
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/common_pitfalls.md:129]: Tagging @t4c1 and @andrjohns can you both look this over to make sure it covers the problem and solution well?

File: doxygen/contributor_help_pages/common_pitfalls.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+When working with arithmetic types, keep in mind that moving Scalars is often less optimal than simply taking their copy. For instance, Stan's `var` type is a PIMPL implimentation, so it simply holds a pointer of size 8 bytes. A `double` is also 8 bytes which just so happens to fit exactly in a [word](https://en.wikipedia.org/wiki/Word_(computer_architecture)) of most modern CPUs. While a reference to a double is also 8 bytes, unless the function is inlined by the compiler, the computer will have to place the reference into cache, then go fetch the value that is being referenced which now takes up two words instead of one!
+
+The general rules to follow for passing values to a function are:
+
+1. If a type is smaller than two words (16 bytes), pass it by value
+2. If you are writing a function for reverse mode, pass values by `const&`
+3. In prim, if you are confident and working with larger types, use perfect forwarding to pass values that can be moved from. Otherwise simply pass values by `const&`.
+
+### Copying non-arena variables to lambdas used in the reverse pass
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/common_pitfalls.md:214]: @bbbales2 can you add make_callback_ptr() stuff here?

File: doxygen/contributor_help_pages/common_pitfalls.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+### `forward_as`
+
+### Nested Stacks
+
+### Scoped Stacks
+
+As an aside, in order to get the actual vector result, Eigen defines the
+`.eval()` operator, so `c.eval()` would return an actual `Eigen::VectorXd`.
+Similarly, Math defines the `eval` function which will evaluate an Eigen
+expression and forward anything that is not an Eigen expression.
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/common_pitfalls.md:237]: @wds15 would you have time to write about Stan's threading scheme for the allocators? I think that would be best to put here

File: doxygen/contributor_help_pages/getting_started.md [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+  for (int i = 0; i < x.size(); ++i) {
+    sum += x_ref.coeff(i) * x_ref.coeff(i);
+  }
+  return sum;
+}
+```
+
+TL;DR: Use Stan's `require` type trait library for limiting a function's scope to certain types.
+
+Reading from left to right, `require_not_st_var_t` requires that a signature's template does _not_ have a _scalar type_ that is Stan's `var` autodiff type. This line is exactly the same as
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/getting_started.md:171]: [todo: me] add link to requires docs

File: doxygen/contributor_help_pages/getting_started.md [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+    arena_v.adj() += (2.0 * vi.adj()) * arena_v.val();
+  });
+}
+```
+
+`make_callback_var(x, f)` is similar to `reverse_pass_callback`, but it constructs
+an autodiff type from the argument `x`. The function `f` is called on the
+reverse pass similarly to `reverse_pass_callback`.
+
+## Argument Types
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/getting_started.md:297]: Maybe everything here should go in another page?

File: doxygen/contributor_help_pages/getting_started.md [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+### Values
+
+`value_of(x)` returns the values of the variable `x`. If `x` is not an
+autodiff type, it simply forwards `x` along. The values of `x` have
+the same shape as `x`. For reverse mode autodiff, values are `double`
+variables. For higher order autodiff (`fvar<var>`, etc.) this is not
+always the case. See `value_of_rec` for a different behavior.
+
+### Values and adjoint extensions to Eigen
+
```
Comment (SteveBronder) [inline doxygen/contributor_help_pages/getting_started.md:537]: This should probs go somewhere else as well

**Comment by SteveBronder** [comment]
I'm going to try to restructure this a lil bit tmrw. I'd like to better structure this like Eigen's docs

File: doxygen/contributor_help_pages/common_pitfalls.md (hunk 1 above)
Comment (t4c1) [reply doxygen/contributor_help_pages/common_pitfalls.md:129]: Looks fine to me.

**Comment by rok-cesnovar** [comment]
Question: Do we want to add which checks to run locally before submitting a PR? Also a note on clang-format and the rest? 
If yes I can handle that.

File: doxygen/contributor_help_pages/common_pitfalls.md (hunk 2 above)
Comment (bbbales2) [reply doxygen/contributor_help_pages/common_pitfalls.md]: Done!

**Comment by SteveBronder** [comment]
Yeah that would be awesome! I think that would go under the testing section

**Comment by SteveBronder** [comment]
@rok-cesnovar if you have time to write up the PR stuff that would be rad but otherwise I think this is good to go!

**Comment by rok-cesnovar** [comment]
Will do over the weekend.

**Comment by rok-cesnovar** [comment]
Obviously that weekend has long passed. Lets merge this.

**Review by rok-cesnovar — APPROVED** [review]

**Pull request merged**